File: CPP/1525.cpp
class Solution {
public:
    // The key point is number of distinct characters in each subarray.
    // Therefore, use frequency table to compare.
    // One vector contains all chars, one starts from none -> update by each char

    int numSplits(string s) {
        vector<int> freA(26,0), freB(26,0);
        int returnVal = 0, index = 0;
        int numA = 0, numB = 0;
        // freB -> num of each char
        // numB -> num of distinct char
        for (char c : s)  {
            freB[c - 'a']++;
            if(freB[c-'a'] == 1) numB++;
        }
        
        for (char c : s){
            index = c - 'a'; 
            freA[index]++;
            freB[index]--;
            if (freA[index] == 1) numA++; // Since index = 1 -> it is a distinct char
            if (freB[index] == 0) numB--; // Since index = 0 -> Lost one distinct charr
            if (numA == numB) returnVal++;
        }
        return returnVal;        
    }
};

File: Python/1525.py
class Solution:
    def numSplits(self, s: str) -> int:
        if len(s) < 2: return 0
        
        listA = [0 for i in range(0,25)] ; listB = [0 for i in range(0,25)]
        numA = 0; numB = 0; returnVal = 0
        base = ord('a')
        for index in s:
            current = ord(index)
            listB[current - base]+=1
            if listB[current - base] == 1: numB += 1
        
        for index in s:
            current = ord(index)
            listA[current - base] += 1
            listB[current - base] -= 1
            if (listA[current - base] == 1): numA += 1
            if (listB[current - base] == 0): numB -= 1
            if (numA == numB): returnVal += 1
        return returnVal
        